net/graphsync_fetcher.go

```diff
@@ -79,37 +89,75 @@ const recursionMultiplier = 4
 //
 // See: https://github.com/filecoin-project/go-filecoin/issues/3175
 func (gsf *GraphSyncFetcher) FetchTipSets(ctx context.Context, tsKey types.TipSetKey, from peer.ID, done func(types.TipSet) (bool, error)) ([]types.TipSet, error) {
-	cids := tsKey.ToSlice()
-	err := gsf.fetchBlocks(ctx, cids, from)
+	rpf := newRequestPeerFinder(gsf.peerTracker, from)
+
+	// fetch initial tipset
+	startingTipset, err := gsf.fetchFirstTipset(ctx, tsKey, rpf)
 	if err != nil {
 		return nil, err
 	}
-	startingTipset, err := gsf.loadTipsetFromCids(ctx, cids)
-	if err != nil {
-		return nil, err
+
+	// fetch remaining tipsets recursively
+	return gsf.fetchRemainingTipsets(ctx, startingTipset, rpf, done)
+}
+
+func (gsf *GraphSyncFetcher) fetchFirstTipset(ctx context.Context, tsKey types.TipSetKey, rpf *requestPeerFinder) (types.TipSet, error) {
+	cids := tsKey.ToSlice()
+	remainingCids := cids
+	for {
+		err := gsf.fetchBlocks(ctx, remainingCids, rpf.CurrentPeer())
+		if err == nil {
+			break
+		}
+		// something went wrong in a graphsync request, but we want to keep trying other peers, so
+		// just log error
+		logGraphsyncFetcher.Infof("Error occurred during Graphsync request: %s", err)
+		remainingCids, err = gsf.missingCids(remainingCids)
```

Comment:
Using these two names together is confusing.  `remaining` means that they remain to be fetched, and `missing` means that they are missing from the set already fetched, so they are actually synonyms here which is pretty unexpected.  You should chose one.  Also you could potentially avoid the rename on line 106, leaving the value as `cids` throughout and filtering it down with your choice of `missingCids` or `remainingCids`.